rfcs/0000-a-forth-userspace.md

```diff
@@ -0,0 +1,72 @@
+# RFC - Forth Userspace
+
+## Goal
+
+In order to make early prototyping and iterating on the design of mnemos **easier**, this RFC proposes the
+addition of a single kind of "userspace" environment, a `forth` vm that can act as a program
+or interactive shell interface. This RFC also proposes a minimal set of capabilities and interfaces
+necessary to make this possible.
+
+## Background
+
+mnemos has some design around what it means to launch kernel services, but has no kind of userspace
+separate from the kernel itself. Designing this "properly" requires a lot of subtle decisions to be
+made, so this rfc proposes to ignore these problems until we are much smarter people later in the
+project.
+
+James has been working on an extensible forth vm written in Rust, and recently Eliza has been helping
+to expand the vm, including the ability to have async host-provided intrinsic functions. While limited,
+this environment is likely suitable for two main things:
+
+* "gluing together syscalls" to test out kernel capabilities and bring up hardware support
+* Crossing the "embedded system"/"computer" rubicon, allowing for runtime developed programs
+
+## The proposal
+
+* The OS will spawn a single userspace entity at boot time
+    * This entity will be a cooperative async task running an interactive forth vm
+    * It will be capable of interacting with the userspace/kernel interface (sending/receiving kernel messages)
+* This "task 0" forth environment will be capable of spawning other userspace tasks
+    * It can only spawn other forth vms, running their own separate environments
+    * When spawning a new task, a snapshot of the current task's dictionary will be made and shared with the child environment
+    * All forth vm tasks will be cooperatively scheduled
+
+## Details
+
+### Changes to `forth3` - the current forth vm
+
+* We'll need to be able to provide a cow-like, reference counted, linked list of previous dictionary fragments
+* We'll need some way of determining "write-like" behaviors that will cause a deepcopy
+    * like obtaining the address of a variable
+    * todo: is this all we need it for? what about variable addrs in variables?
+* We'll need some way of allocating + deepcopying dictionary fragments
+
+### Changes to `mnemos`
+
+* We'll need to add some kind of pattern/interface for managing cooperative userspace tasks
+    * We will probably NOT enforce preemptive multitasking for now
+    * We will probably NOT enforce memory protection/mapping for now
+    * userspace tasks yield when they feel like it, kernel then serves each tasks's executor like a metaexecutor
```

Comment:
not totally sure what you mean by "serves each task's executor like a metaexacutor", exactly --- each forth VM would just get a single task in the kernel's maitake scheduler, for now

Reply:
Yeah, I guess I made two steps here:

At first, the kernel will just run tasks "raw" in kernel space.

Later, when there is real control separation between kernel and userspace, I expect each userspace "process" to have it's own executor. I sorta plan to expose userspace processes as futures in the kernel - but polling them yields control flow to the userspace program - essentially handing control to the subordinate executor, until either it yields (no pending tasks), the outgoing ringbuffer is full, etc.

A fair amount of this is probably worth a whole separate RFC later, but this is where my head was at when writing that.

Reply:
i think that makes sense, when userspace programs can be compiled Rust realcodes as well as interpreted forth programs! i was just thinking in the context of the forth stuff. would love to see your thoughts on the ABI for gluing an async rust binary into the OS' scheduler, eventually!

Reply:
Yeah, I agree that's better to save for another day. I have lots of ideas, some of them might be good. Not strictly necessary now :)